### PR TITLE
fix: BigInt Support for DynamoDB Convert

### PR DESCRIPTION
**Pull request by TechTeam12**
##### Checklist

- [/] `npm run test` passes
- [/] changelog is added, `npm run add-change`

Issue reference:
https://github.com/aws/aws-sdk-js/issues/3018

### CODE REVIEW

**Comment by TechTeam12** [comment]
> # [Codecov](https://codecov.io/gh/aws/aws-sdk-js/pull/3019?src=pr&el=h1) Report
> 
> > Merging [#3019](https://codecov.io/gh/aws/aws-sdk-js/pull/3019?src=pr&el=desc) into [master](https://codecov.io/gh/aws/aws-sdk-js/commit/2c93f42bd7d037d4ef4c6d765c5dd20046b852a0?src=pr&el=desc) will **decrease** coverage by `59.62%`.
> > The diff coverage is `100%`.
> 
> [Impacted file tree graph](https://codecov.io/gh/aws/aws-sdk-js/pull/3019?src=pr&el=tree)
> 
> ```diff
> @@             Coverage Diff             @@
> ##           master    #3019       +/-   ##
> ===========================================
> - Coverage   97.05%   37.42%   -59.63%     
> ===========================================
>   Files         328      328               
>   Lines        9497     9499        +2     
>   Branches     1743     1744        +1     
> ===========================================
> - Hits         9217     3555     -5662     
> - Misses        280     5944     +5664
> ```
> 
> [Impacted Files](https://codecov.io/gh/aws/aws-sdk-js/pull/3019?src=pr&el=tree) 	Coverage Δ 	
> [lib/dynamodb/converter.js](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree#diff-bGliL2R5bmFtb2RiL2NvbnZlcnRlci5qcw==) 	`22.41% <100%> (-76.71%)` 	arrow_down
> [lib/services/sqs.js](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree#diff-bGliL3NlcnZpY2VzL3Nxcy5qcw==) 	`3.38% <0%> (-96.62%)` 	arrow_down
> [lib/param_validator.js](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree#diff-bGliL3BhcmFtX3ZhbGlkYXRvci5qcw==) 	`1.38% <0%> (-95.14%)` 	arrow_down
> [lib/signers/v4.js](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree#diff-bGliL3NpZ25lcnMvdjQuanM=) 	`5.45% <0%> (-94.55%)` 	arrow_down
> [lib/services/s3control.js](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree#diff-bGliL3NlcnZpY2VzL3MzY29udHJvbC5qcw==) 	`5.71% <0%> (-94.29%)` 	arrow_down
> [lib/services/s3.js](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree#diff-bGliL3NlcnZpY2VzL3MzLmpz) 	`3.97% <0%> (-94.23%)` 	arrow_down
> [lib/credentials/remote_credentials.js](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree#diff-bGliL2NyZWRlbnRpYWxzL3JlbW90ZV9jcmVkZW50aWFscy5qcw==) 	`2.89% <0%> (-94.21%)` 	arrow_down
> [lib/signers/s3.js](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree#diff-bGliL3NpZ25lcnMvczMuanM=) 	`6.06% <0%> (-93.94%)` 	arrow_down
> [lib/services/cloudsearchdomain.js](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree#diff-bGliL3NlcnZpY2VzL2Nsb3Vkc2VhcmNoZG9tYWluLmpz) 	`6.66% <0%> (-93.34%)` 	arrow_down
> [lib/services/glacier.js](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree#diff-bGliL3NlcnZpY2VzL2dsYWNpZXIuanM=) 	`4.54% <0%> (-93.19%)` 	arrow_down
> ... and [305 more](https://codecov.io/gh/aws/aws-sdk-js/pull/3019/diff?src=pr&el=tree-more) 		
> 
> [Continue to review full report at Codecov](https://codecov.io/gh/aws/aws-sdk-js/pull/3019?src=pr&el=continue).
> 
> > **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> > `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> > Powered by [Codecov](https://codecov.io/gh/aws/aws-sdk-js/pull/3019?src=pr&el=footer). Last update [2c93f42...631e33e](https://codecov.io/gh/aws/aws-sdk-js/pull/3019?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by TechTeam12** [comment]
Accidentally closed this, was trying to cancel on a comment.

**Comment by ajredniwja** [comment]
Hey @TechTeam12,

I was able to test it and it works without breaking anything, need @AllanFly120 to review it before merging it.

**Comment by ajredniwja** [comment]
I also the description from ```bigfix``` to a ```feature```.

**Comment by TechTeam12** [comment]
> Hey @TechTeam12,
> 
> I was able to test it and it works without breaking anything, need @AllanFly120 to review it before merging it.

Woo, woo. 

Hopefully this gets merged soon. Would really love to finish off this project I'm working before I leave for the holidays.

**Comment by TechTeam12** [comment]
@ajredniwja or @AllanFly120 
Any update on this?

**Comment by TechTeam12** [comment]
Hey I know I timed this poorly with the holidays, but any update would be really appreciated. I'd very much like to take advantage of the BigInt data types to store my time stamps into dynamo. Please let me know where we are, and if there is something else that needs to be done.

**Comment by TechTeam12** [comment]
I've updated this branch to have the latest code from master.

**Comment by TechTeam12** [comment]
Updated to be current to master.

**Comment by ajredniwja** [comment]
I was able to confirm that it doesn't break anything when inserting any BigInt value.

**Comment by AllanFly120** [comment]
@TechTeam12 How about deserializing BigInt from response? Can you add a unit test for it? Do you need to to handle BigInt specifically [here](https://github.com/aws/aws-sdk-js/pull/3019/files#diff-c047b818394aa57b0b37173d285ee689R226)?

**Comment by TechTeam12** [comment]
@AllanFly120 I'll get it done this evening. Thanks!

File: lib/dynamodb/converter.js [hunk 1]
```diff
@@ -221,7 +223,16 @@ function formatList(data, options) {
...
  */
 function convertNumber(value, wrapNumbers) {
-  return wrapNumbers ? new NumberValue(value) : Number(value);
+  if (wrapNumbers) {
+    return new NumberValue(value);
+  }
+  try {
+    return Number.isSafeInteger(value) ? Number(value) : BigInt(value);
+  } catch (err) {
+    if (err === 'BigInt is not defined') {
```
Comment (TechTeam12) [inline lib/dynamodb/converter.js]: @AllanFly120 Do you have any way that the SDK could throw a warning to the user that the number they're deserializing is not a safe int, and to use the wrap option?

Comment (TechTeam12) [reply lib/dynamodb/converter.js]: Ignore the BigInt error check. This is not valid. 

But still curious, does the AWS SDK throw any warning messages? I noticed before if you did not provide the wrapNumbers option that you could actually get an unsafe number back? 
his is interesting. Say, I used the UI and added a Number to the dynamo table. If I queried the table and did not pass that option it looks like it'd return an unsafe casted number of that value and not warn the user of that. Personally,as a consumer of this SDK I'd appreciate some kind of warning and prompting to use that option if this happened.

**Comment by TechTeam12** [comment]
@AllanFly120 Any update on this?

File: lib/dynamodb/converter.js (hunk 1 above)
Comment (TechTeam12) [reply lib/dynamodb/converter.js]: @AllanFly120

**Comment by TechTeam12** [comment]
@AllanFly120 Any word?

**Comment by ajredniwja** [comment]
> @AllanFly120 Any word?

Allan wont be available until next week, asked @trivikr to review the code changes.

**Comment by TechTeam12** [comment]
Awesome, thank you @ajredniwja

File: lib/dynamodb/converter.js (hunk 1 above)
Comment (trivikr) [reply lib/dynamodb/converter.js]: @TechTeam12 Can you create a separate issue for discussing this?

**Comment by trivikr** [comment]
Merged.
It'll be released with v2.610.0 around `Fri Jan 24 2020 12:00:00 GMT-0800 (Pacific Standard Time)`

File: lib/dynamodb/converter.js (hunk 1 above)
Comment (TechTeam12) [reply lib/dynamodb/converter.js]: Will do!

Comment (TechTeam12) [reply lib/dynamodb/converter.js]: Little long winded, but: 
https://github.com/aws/aws-sdk-js/issues/3065

**Pull request merged**